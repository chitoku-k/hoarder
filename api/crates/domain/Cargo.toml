[package]
name = "domain"
authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
doctest = false

[dependencies.chrono]
version = "0.4.39"
features = ["clock", "std"]
default-features = false

[dependencies.derive_more]
version = "2.0.1"
features = ["constructor", "deref", "display", "from"]

[dependencies.futures]
version = "0.3.31"
features = ["std"]
default-features = false

[dependencies.indenter]
version = "0.3.3"

[dependencies.ordermap]
version = "0.5.5"

[dependencies.percent-encoding]
version = "2.3.1"

[dependencies.regex]
version = "1.11.1"

[dependencies.serde]
version = "1.0.218"
features = ["derive"]

[dependencies.strum]
version = "0.27.0"
features = ["derive"]

[dependencies.thiserror]
version = "2.0.11"

[dependencies.tokio]
version = "1.43.0"
features = ["macros", "rt-multi-thread"]

[dependencies.tracing]
version = "0.1.41"

[dependencies.uuid]
version = "1.13.1"
features = ["serde", "v4"]

[dev-dependencies.anyhow]
version = "1.0.95"

[dev-dependencies.indoc]
version = "2.0.5"

[dev-dependencies.mockall]
version = "0.13.1"

[dev-dependencies.pretty_assertions]
version = "1.4.1"
features = ["unstable"]

[dev-dependencies.serial_test]
version = "3.2.0"
