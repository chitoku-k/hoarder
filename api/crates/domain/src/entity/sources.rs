use chrono::{DateTime, Utc};
use derive_more::{Deref, Display, From};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{entity::external_services::{ExternalMetadata, ExternalService}, error::{ErrorKind, Result}};

#[derive(Clone, Copy, Debug, Default, Deref, Deserialize, Display, Eq, From, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct SourceId(Uuid);

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Source {
    pub id: SourceId,
    pub external_service: ExternalService,
    pub external_metadata: ExternalMetadata,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Source {
    pub fn url(&self) -> Option<String> {
        self.external_metadata.url(self.external_service.base_url.as_deref())
    }

    pub fn validate(&self) -> Result<()> {
        if self.external_metadata.kind().is_none_or(|kind| kind == self.external_service.kind) {
            Ok(())
        } else {
            Err(ErrorKind::SourceMetadataNotMatch { kind: self.external_service.kind.clone() })?
        }
    }
}
