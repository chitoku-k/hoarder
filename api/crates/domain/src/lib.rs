#![allow(clippy::too_many_arguments)]

pub mod entity;
pub mod parser;
pub mod processor;
pub mod repository;
pub mod service;
