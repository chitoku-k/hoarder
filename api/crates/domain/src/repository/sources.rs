use async_trait::async_trait;

use crate::{
    entity::{
        external_services::{ExternalMetadata, ExternalServiceId},
        sources::{Source, SourceId},
    },
    repository::DeleteResult,
};

#[cfg_attr(test, mockall::automock)]
#[async_trait]
pub trait SourcesRepository: Send + Sync + 'static {
    /// Creates a source.
    async fn create(&self, external_service_id: ExternalServiceId, external_metadata: ExternalMetadata) -> anyhow::Result<Source>;

    /// Fetches the source by its external metadata.
    async fn fetch_by_external_metadata(&self, external_service_id: ExternalServiceId, external_metadata: ExternalMetadata) -> anyhow::Result<Source>;

    /// Updates the source by ID.
    async fn update_by_id(&self, id: SourceId, external_service_id: Option<ExternalServiceId>, external_metadata: Option<ExternalMetadata>) -> anyhow::Result<Source>;

    /// Deletes the source by ID.
    async fn delete_by_id(&self, id: SourceId) -> anyhow::Result<DeleteResult>;
}
