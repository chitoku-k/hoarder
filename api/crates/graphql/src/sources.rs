use async_graphql::{InputObject, OneofObject, SimpleObject};
use chrono::{DateTime, Utc};
use domain::entity::{external_services, sources};
use serde::Serialize;
use uuid::Uuid;

use crate::{
    error::ErrorKind,
    external_services::ExternalService,
};

#[derive(SimpleObject)]
pub(crate) struct Source {
    id: Uuid,
    external_service: ExternalService,
    external_metadata: serde_json::Value,
    created_at: DateTime<Utc>,
    updated_at: DateTime<Utc>,
}

#[derive(Debug, Eq, OneofObject, PartialEq, Serialize)]
#[graphql(name = "ExternalMetadataInput")]
#[serde(rename_all = "snake_case")]
pub(crate) enum ExternalMetadata {
    Bluesky(ExternalMetadataIdCreatorId),
    Fantia(ExternalMetadataId),
    Mastodon(ExternalMetadataIdOptionalCreatorId),
    Misskey(ExternalMetadataId),
    Nijie(ExternalMetadataId),
    Pixiv(ExternalMetadataId),
    #[graphql(name = "pixiv_fanbox")]
    PixivFanbox(ExternalMetadataIdCreatorId),
    Pleroma(ExternalMetadataId),
    Seiga(ExternalMetadataId),
    Skeb(ExternalMetadataIdCreatorId),
    Threads(ExternalMetadataIdOptionalCreatorId),
    Twitter(ExternalMetadataIdOptionalCreatorId),
    Website(ExternalMetadataUrl),
    Custom(serde_json::Value),
}

#[derive(Debug, Eq, InputObject, PartialEq, Serialize)]
#[graphql(name = "ExternalMetadataIdInput")]
pub(crate) struct ExternalMetadataId {
    id: String,
}

#[derive(Debug, Eq, InputObject, PartialEq, Serialize)]
#[graphql(name = "ExternalMetadataIdCreatorIdInput")]
#[serde(rename_all = "camelCase")]
pub(crate) struct ExternalMetadataIdCreatorId {
    id: String,
    creator_id: String,
}

#[derive(Debug, Eq, InputObject, PartialEq, Serialize)]
#[graphql(name = "ExternalMetadataIdOptionalCreatorIdInput")]
#[serde(rename_all = "camelCase")]
pub(crate) struct ExternalMetadataIdOptionalCreatorId {
    id: String,
    creator_id: Option<String>,
}

#[derive(Debug, Eq, InputObject, PartialEq, Serialize)]
#[graphql(name = "ExternalMetadataUrlInput")]
pub(crate) struct ExternalMetadataUrl {
    url: String,
}

impl TryFrom<external_services::ExternalMetadata> for ExternalMetadata {
    type Error = ErrorKind;

    fn try_from(value: external_services::ExternalMetadata) -> Result<Self, Self::Error> {
        use external_services::ExternalMetadata::*;
        match value {
            Bluesky { id, creator_id } => Ok(Self::Bluesky(ExternalMetadataIdCreatorId { id, creator_id })),
            Fantia { id } => Ok(Self::Fantia(ExternalMetadataId { id: id.to_string() })),
            Mastodon { id, creator_id } => Ok(Self::Mastodon(ExternalMetadataIdOptionalCreatorId { id: id.to_string(), creator_id })),
            Misskey { id } => Ok(Self::Misskey(ExternalMetadataId { id })),
            Nijie { id } => Ok(Self::Nijie(ExternalMetadataId { id: id.to_string() })),
            Pixiv { id } => Ok(Self::Pixiv(ExternalMetadataId { id: id.to_string() })),
            PixivFanbox { id, creator_id } => Ok(Self::PixivFanbox(ExternalMetadataIdCreatorId { id: id.to_string(), creator_id })),
            Pleroma { id } => Ok(Self::Pleroma(ExternalMetadataId { id })),
            Seiga { id } => Ok(Self::Seiga(ExternalMetadataId { id: id.to_string() })),
            Skeb { id, creator_id } => Ok(Self::Skeb(ExternalMetadataIdCreatorId { id: id.to_string(), creator_id })),
            Threads { id, creator_id } => Ok(Self::Threads(ExternalMetadataIdOptionalCreatorId { id, creator_id })),
            Twitter { id, creator_id } => Ok(Self::Twitter(ExternalMetadataIdOptionalCreatorId { id: id.to_string(), creator_id })),
            Website { url } => Ok(Self::Website(ExternalMetadataUrl { url })),
            Custom(v) => Ok(Self::Custom(serde_json::from_str(&v).map_err(|_| ErrorKind::SourceMetadataInvalid)?)),
        }
    }
}

impl TryFrom<ExternalMetadata> for external_services::ExternalMetadata {
    type Error = ErrorKind;

    fn try_from(value: ExternalMetadata) -> Result<Self, Self::Error> {
        use ExternalMetadata::*;
        match value {
            Bluesky(ExternalMetadataIdCreatorId { id, creator_id }) => Ok(Self::Bluesky { id, creator_id }),
            Fantia(ExternalMetadataId { id }) => Ok(Self::Fantia { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)? }),
            Mastodon(ExternalMetadataIdOptionalCreatorId { id, creator_id }) => Ok(Self::Mastodon { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)?, creator_id }),
            Misskey(ExternalMetadataId { id }) => Ok(Self::Misskey { id }),
            Nijie(ExternalMetadataId { id }) => Ok(Self::Nijie { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)? }),
            Pixiv(ExternalMetadataId { id }) => Ok(Self::Pixiv { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)? }),
            PixivFanbox(ExternalMetadataIdCreatorId { id, creator_id }) => Ok(Self::PixivFanbox { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)?, creator_id }),
            Pleroma(ExternalMetadataId { id }) => Ok(Self::Pleroma { id }),
            Seiga(ExternalMetadataId { id }) => Ok(Self::Seiga { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)? }),
            Skeb(ExternalMetadataIdCreatorId { id, creator_id }) => Ok(Self::Skeb { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)?, creator_id }),
            Threads(ExternalMetadataIdOptionalCreatorId { id, creator_id }) => Ok(Self::Threads { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)?, creator_id }),
            Twitter(ExternalMetadataIdOptionalCreatorId { id, creator_id }) => Ok(Self::Twitter { id: id.parse().map_err(|_| ErrorKind::SourceMetadataInvalid)?, creator_id }),
            Website(ExternalMetadataUrl { url }) => Ok(Self::Website { url }),
            Custom(v) => Ok(Self::Custom(v.to_string())),
        }
    }
}

impl TryFrom<sources::Source> for Source {
    type Error = ErrorKind;

    fn try_from(source: sources::Source) -> Result<Self, Self::Error> {
        let external_metadata = ExternalMetadata::try_from(source.external_metadata)?;

        Ok(Self {
            id: *source.id,
            external_service: source.external_service.into(),
            external_metadata: serde_json::to_value(external_metadata).map_err(|_| ErrorKind::SourceMetadataInvalid)?,
            created_at: source.created_at,
            updated_at: source.updated_at,
        })
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;
    use serde_json::json;

    use super::*;

    #[test]
    fn convert_bluesky() {
        let metadata = ExternalMetadata::Bluesky(ExternalMetadataIdCreatorId { id: "abcdefghi".to_string(), creator_id: "creator_01".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Bluesky { id: "abcdefghi".to_string(), creator_id: "creator_01".to_string() });

        let metadata = external_services::ExternalMetadata::Bluesky { id: "abcdefghi".to_string(), creator_id: "creator_01".to_string() };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Bluesky(ExternalMetadataIdCreatorId { id: "abcdefghi".to_string(), creator_id: "creator_01".to_string() }));
    }

    #[test]
    fn convert_fantia() {
        let metadata = ExternalMetadata::Fantia(ExternalMetadataId { id: "123456789".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Fantia { id: 123456789 });

        let metadata = external_services::ExternalMetadata::Fantia { id: 123456789 };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Fantia(ExternalMetadataId { id: "123456789".to_string() }));
    }

    #[test]
    fn convert_mastodon() {
        let metadata = ExternalMetadata::Mastodon(ExternalMetadataIdOptionalCreatorId { id: "123456789".to_string(), creator_id: Some("creator_01".to_string()) });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Mastodon { id: 123456789, creator_id: Some("creator_01".to_string()) });

        let metadata = external_services::ExternalMetadata::Mastodon { id: 123456789, creator_id: Some("creator_01".to_string()) };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Mastodon(ExternalMetadataIdOptionalCreatorId { id: "123456789".to_string(), creator_id: Some("creator_01".to_string()) }));
    }

    #[test]
    fn convert_misskey() {
        let metadata = ExternalMetadata::Misskey(ExternalMetadataId { id: "abcdefghi".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Misskey { id: "abcdefghi".to_string() });

        let metadata = external_services::ExternalMetadata::Misskey { id: "abcdefghi".to_string() };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Misskey(ExternalMetadataId { id: "abcdefghi".to_string() }));
    }

    #[test]
    fn convert_nijie() {
        let metadata = ExternalMetadata::Nijie(ExternalMetadataId { id: "123456789".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Nijie { id: 123456789 });

        let metadata = external_services::ExternalMetadata::Nijie { id: 123456789 };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Nijie(ExternalMetadataId { id: "123456789".to_string() }));
    }

    #[test]
    fn convert_pixiv() {
        let metadata = ExternalMetadata::Pixiv(ExternalMetadataId { id: "123456789".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Pixiv { id: 123456789 });

        let metadata = external_services::ExternalMetadata::Pixiv { id: 123456789 };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Pixiv(ExternalMetadataId { id: "123456789".to_string() }));
    }

    #[test]
    fn convert_pixiv_fanbox() {
        let metadata = ExternalMetadata::PixivFanbox(ExternalMetadataIdCreatorId { id: "123456789".to_string(), creator_id: "creator_01".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::PixivFanbox { id: 123456789, creator_id: "creator_01".to_string() });

        let metadata = external_services::ExternalMetadata::PixivFanbox { id: 123456789, creator_id: "creator_01".to_string() };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::PixivFanbox(ExternalMetadataIdCreatorId { id: "123456789".to_string(), creator_id: "creator_01".to_string() }));
    }

    #[test]
    fn convert_pleroma() {
        let metadata = ExternalMetadata::Pleroma(ExternalMetadataId { id: "abcdefghi".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Pleroma { id: "abcdefghi".to_string() });

        let metadata = external_services::ExternalMetadata::Pleroma { id: "abcdefghi".to_string() };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Pleroma(ExternalMetadataId { id: "abcdefghi".to_string() }));
    }

    #[test]
    fn convert_seiga() {
        let metadata = ExternalMetadata::Seiga(ExternalMetadataId { id: "123456789".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Seiga { id: 123456789 });

        let metadata = external_services::ExternalMetadata::Seiga { id: 123456789 };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Seiga(ExternalMetadataId { id: "123456789".to_string() }));
    }

    #[test]
    fn convert_skeb() {
        let metadata = ExternalMetadata::Skeb(ExternalMetadataIdCreatorId { id: "123456789".to_string(), creator_id: "creator_01".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Skeb { id: 123456789, creator_id: "creator_01".to_string() });

        let metadata = external_services::ExternalMetadata::Skeb { id: 123456789, creator_id: "creator_01".to_string() };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Skeb(ExternalMetadataIdCreatorId { id: "123456789".to_string(), creator_id: "creator_01".to_string() }));
    }

    #[test]
    fn convert_threads() {
        let metadata = ExternalMetadata::Threads(ExternalMetadataIdOptionalCreatorId { id: "abcdefghi".to_string(), creator_id: Some("creator_01".to_string()) });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Threads { id: "abcdefghi".to_string(), creator_id: Some("creator_01".to_string()) });

        let metadata = external_services::ExternalMetadata::Threads { id: "abcdefghi".to_string(), creator_id: Some("creator_01".to_string()) };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Threads(ExternalMetadataIdOptionalCreatorId { id: "abcdefghi".to_string(), creator_id: Some("creator_01".to_string()) }));
    }

    #[test]
    fn convert_twitter() {
        let metadata = ExternalMetadata::Twitter(ExternalMetadataIdOptionalCreatorId { id: "123456789".to_string(), creator_id: Some("creator_01".to_string()) });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Twitter { id: 123456789, creator_id: Some("creator_01".to_string()) });

        let metadata = external_services::ExternalMetadata::Twitter { id: 123456789, creator_id: Some("creator_01".to_string()) };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Twitter(ExternalMetadataIdOptionalCreatorId { id: "123456789".to_string(), creator_id: Some("creator_01".to_string()) }));
    }

    #[test]
    fn convert_website() {
        let metadata = ExternalMetadata::Website(ExternalMetadataUrl { url: "https://example.com".to_string() });
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Website { url: "https://example.com".to_string() });

        let metadata = external_services::ExternalMetadata::Website { url: "https://example.com".to_string() };
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Website(ExternalMetadataUrl { url: "https://example.com".to_string() }));
    }

    #[test]
    fn convert_custom() {
        let metadata = ExternalMetadata::Custom(json!({ "id": 123456789 }));
        let actual = external_services::ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, external_services::ExternalMetadata::Custom(r#"{"id":123456789}"#.to_string()));

        let metadata = external_services::ExternalMetadata::Custom(r#"{"id":123456789}"#.to_string());
        let actual = ExternalMetadata::try_from(metadata).unwrap();

        assert_eq!(actual, ExternalMetadata::Custom(json!({ "id": 123456789 })));
    }
}
