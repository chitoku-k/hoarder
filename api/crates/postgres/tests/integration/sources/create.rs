use domain::{
    entity::external_services::{ExternalMetadata, ExternalService, ExternalServiceId, ExternalServiceKind},
    repository::sources::SourcesRepository,
};
use postgres::sources::PostgresSourcesRepository;
use pretty_assertions::assert_eq;
use serde_json::json;
use sqlx::Row;
use test_context::test_context;
use uuid::{uuid, Uuid};

use super::DatabaseContext;

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds_with_default(ctx: &DatabaseContext) {
    let repository = PostgresSourcesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        ExternalServiceId::from(uuid!("4e0c68c7-e5ec-4d60-b9eb-733f47290cd3")),
        ExternalMetadata::Pixiv { id: 123456789 },
    ).await.unwrap();

    assert_eq!(
        actual.external_service,
        ExternalService {
            id: ExternalServiceId::from(uuid!("4e0c68c7-e5ec-4d60-b9eb-733f47290cd3")),
            slug: "pixiv".to_string(),
            kind: ExternalServiceKind::Pixiv,
            name: "pixiv".to_string(),
            base_url: Some("https://www.pixiv.net".to_string()),
            url_pattern: Some(r"^https?://www\.pixiv\.net/(?:artworks/|member_illust\.php\?(?:|.+&)illust_id=)(?<id>\d+)(?:[?&#].*)?$".to_string()),
        },
    );
    assert_eq!(actual.external_metadata, ExternalMetadata::Pixiv { id: 123456789 });

    let actual = sqlx::query(r#"SELECT "id", "external_service_id", "external_metadata", "external_metadata_extra" FROM "sources" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<Uuid, &str>("external_service_id"), uuid!("4e0c68c7-e5ec-4d60-b9eb-733f47290cd3"));
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata"),
        json!({
            "type": "pixiv",
            "id": 123456789,
        }),
    );
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata_extra"),
        json!({
            "type": "pixiv",
        }),
    );
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds_with_extra(ctx: &DatabaseContext) {
    let repository = PostgresSourcesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        ExternalServiceId::from(uuid!("99a9f0e8-1097-4b7f-94f2-2a7d2cc786ab")),
        ExternalMetadata::X{ id: 123456789, creator_id: Some("creator_01".to_string()) },
    ).await.unwrap();

    assert_eq!(
        actual.external_service,
        ExternalService {
            id: ExternalServiceId::from(uuid!("99a9f0e8-1097-4b7f-94f2-2a7d2cc786ab")),
            slug: "x".to_string(),
            kind: ExternalServiceKind::X,
            name: "X".to_string(),
            base_url: Some("https://x.com".to_string()),
            url_pattern: Some(r"^https?://(?:twitter\.com|x\.com)/(?<creatorId>[^/]+)/status/(?<id>\d+)(?:[/?#].*)?$".to_string()),
        },
    );
    assert_eq!(actual.external_metadata, ExternalMetadata::X { id: 123456789, creator_id: Some("creator_01".to_string()) });

    let actual = sqlx::query(r#"SELECT "id", "external_service_id", "external_metadata", "external_metadata_extra" FROM "sources" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<Uuid, &str>("external_service_id"), uuid!("99a9f0e8-1097-4b7f-94f2-2a7d2cc786ab"));
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata"),
        json!({
            "type": "x",
            "id": 123456789,
        }),
    );
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata_extra"),
        json!({
            "type": "x",
            "creatorId": "creator_01",
        }),
    );
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds_with_custom_object(ctx: &DatabaseContext) {
    let repository = PostgresSourcesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        ExternalServiceId::from(uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14")),
        ExternalMetadata::Custom(r#"{"id":123456789}"#.to_string()),
    ).await.unwrap();

    assert_eq!(
        actual.external_service,
        ExternalService {
            id: ExternalServiceId::from(uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14")),
            slug: "whatever".to_string(),
            kind: ExternalServiceKind::Custom("custom".to_string()),
            name: "Custom".to_string(),
            base_url: None,
            url_pattern: None,
        },
    );
    assert_eq!(actual.external_metadata, ExternalMetadata::Custom(r#"{"id":123456789}"#.to_string()));

    let actual = sqlx::query(r#"SELECT "id", "external_service_id", "external_metadata", "external_metadata_extra" FROM "sources" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<Uuid, &str>("external_service_id"), uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14"));
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata"),
        json!({
            "id": 123456789,
        }),
    );
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata_extra"),
        json!({}),
    );
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds_with_custom_string(ctx: &DatabaseContext) {
    let repository = PostgresSourcesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        ExternalServiceId::from(uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14")),
        ExternalMetadata::Custom(r#""123456789abcdefg""#.to_string()),
    ).await.unwrap();

    assert_eq!(
        actual.external_service,
        ExternalService {
            id: ExternalServiceId::from(uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14")),
            slug: "whatever".to_string(),
            kind: ExternalServiceKind::Custom("custom".to_string()),
            name: "Custom".to_string(),
            base_url: None,
            url_pattern: None,
        },
    );
    assert_eq!(actual.external_metadata, ExternalMetadata::Custom(r#""123456789abcdefg""#.to_string()));

    let actual = sqlx::query(r#"SELECT "id", "external_service_id", "external_metadata", "external_metadata_extra" FROM "sources" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<Uuid, &str>("external_service_id"), uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14"));
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata"),
        json!("123456789abcdefg"),
    );
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata_extra"),
        json!({}),
    );
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds_with_custom_number(ctx: &DatabaseContext) {
    let repository = PostgresSourcesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        ExternalServiceId::from(uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14")),
        ExternalMetadata::Custom("123456789".to_string()),
    ).await.unwrap();

    assert_eq!(
        actual.external_service,
        ExternalService {
            id: ExternalServiceId::from(uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14")),
            slug: "whatever".to_string(),
            kind: ExternalServiceKind::Custom("custom".to_string()),
            name: "Custom".to_string(),
            base_url: None,
            url_pattern: None,
        },
    );
    assert_eq!(actual.external_metadata, ExternalMetadata::Custom("123456789".to_string()));

    let actual = sqlx::query(r#"SELECT "id", "external_service_id", "external_metadata", "external_metadata_extra" FROM "sources" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<Uuid, &str>("external_service_id"), uuid!("6c07eb4d-93a1-4efd-afce-e13f8f2c0e14"));
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata"),
        json!(123456789),
    );
    assert_eq!(
        actual.get::<serde_json::Value, &str>("external_metadata_extra"),
        json!({}),
    );
}
