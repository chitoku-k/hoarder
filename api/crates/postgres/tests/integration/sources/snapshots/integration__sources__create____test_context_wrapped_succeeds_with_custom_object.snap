---
source: crates/postgres/tests/integration/sources/create.rs
expression: ctx.queries()
---
[[queries]]
sql = '''
SELECT
  "id",
  "slug",
  "kind",
  "name",
  "base_url",
  "url_pattern"
FROM
  "external_services"
WHERE
  "id" = $1'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = '''
INSERT INTO
  "sources" (
    "external_service_id",
    "external_metadata",
    "external_metadata_extra"
  )
VALUES
  ($1, $2, $3)
RETURNING
  "id",
  "external_service_id",
  "external_metadata",
  "external_metadata_extra",
  "created_at",
  "updated_at"'''
rows_affected = 1
rows_returned = 1

[[queries]]
sql = 'COMMIT'
rows_affected = 0
rows_returned = 0
